File: backend/arrlink/auth/lockout.py
"""Password-login lockout: a flat lockout after too many failed attempts."""
from __future__ import annotations

import time
from typing import NamedTuple

from ..state import State

# Flat lockout, not exponential backoff -- a single-account model doesn't
# benefit from that complexity. 5 failures within a 15-minute window locks
# the account for 15 minutes. Hardcoded, not Settings-editable: anyone who
# can already reach PUT /api/settings shouldn't be able to raise or disable
# the very control meant to slow down a credential-guessing client hitting
# POST /api/auth/password.
THRESHOLD = 5
WINDOW_S = 15 * 60.0
LOCKOUT_S = 15 * 60.0


class LockoutStatus(NamedTuple):
    locked: bool
    locked_until: float | None
    fail_count: int


def _row(db: State, username: str) -> dict | None:
    row = db.query_one(
        "SELECT * FROM login_attempts WHERE username=?", (username,)
    )
    return dict(row) if row else None


def status(db: State, username: str) -> LockoutStatus:
    row = _row(db, username)
    if row is None:
        return LockoutStatus(False, None, 0)
    locked_until = row["locked_until"]
    locked = bool(locked_until and locked_until > time.time())
    return LockoutStatus(locked, locked_until if locked else None, row["fail_count"])


def record_failure(db: State, username: str) -> LockoutStatus:
    """Record one failed attempt. No-op on the counters while already
    locked (doesn't extend the lockout window on repeated guesses against a
    locked account)."""
    now = time.time()
    row = _row(db, username)
    was_locked = bool(row and row["locked_until"] and row["locked_until"] > now)

    if was_locked:
        return LockoutStatus(True, row["locked_until"], row["fail_count"])

    if row is None or not row["first_fail_at"] or now - row["first_fail_at"] > WINDOW_S:
        # No row yet, or the previous window (or a since-expired lockout's
        # window) has passed: start fresh.
        fail_count = 1
        first_fail_at = now
    else:
        fail_count = row["fail_count"] + 1
        first_fail_at = row["first_fail_at"]

    locked_until = now + LOCKOUT_S if fail_count >= THRESHOLD else None

    db.execute(
        "INSERT INTO login_attempts (username, fail_count, first_fail_at, "
        "last_fail_at, locked_until) VALUES (?,?,?,?,?) "
        "ON CONFLICT(username) DO UPDATE SET fail_count=excluded.fail_count, "
        "first_fail_at=excluded.first_fail_at, last_fail_at=excluded.last_fail_at, "
        "locked_until=excluded.locked_until",
        (username, fail_count, first_fail_at, now, locked_until),
    )
    db.commit()

    if locked_until is not None:
        # was_locked is already known False here, so any locked_until we
        # just set is a fresh transition -- log once, not on every
        # subsequent attempt during the lockout (those hit the early
        # return above instead).
        db.log_event(
            "warn",
            f"password login locked for {LOCKOUT_S / 60:.0f} min after "
            f"{fail_count} failed attempts (username={username!r})",
        )

    return LockoutStatus(locked_until is not None, locked_until, fail_count)


def reset(db: State, username: str) -> None:
    db.execute("DELETE FROM login_attempts WHERE username=?", (username,))
    db.commit()

File: backend-ts/src/auth/lockout.ts
import { eq } from 'drizzle-orm'
import type { DbClient } from '../db/client.js'
import { logEvent } from '../db/events.js'
import { loginAttempts } from '../db/schema.js'

// Flat lockout: 5 failures within 15 minutes locks the account for 15
// minutes. Hardcoded, not Settings-editable.
export const THRESHOLD = 5
export const WINDOW_S = 15 * 60
export const LOCKOUT_S = 15 * 60

export interface LockoutStatus {
  locked: boolean
  lockedUntil: number | null
  failCount: number
}

// Accepts either the top-level DbClient or a transaction's `tx` handle --
// drizzle types them differently, but both expose what this module needs.
type Queryable = Pick<DbClient, 'select' | 'insert' | 'delete'>

function row(db: Queryable, username: string) {
  return db.select().from(loginAttempts).where(eq(loginAttempts.username, username)).get()
}

export function status(db: DbClient, username: string): LockoutStatus {
  const r = row(db, username)
  if (!r) return { locked: false, lockedUntil: null, failCount: 0 }
  const now = Date.now() / 1000
  const locked = Boolean(r.lockedUntil && r.lockedUntil > now)
  return { locked, lockedUntil: locked ? r.lockedUntil : null, failCount: r.failCount }
}

/** Record one failed attempt, atomically. better-sqlite3 is single-connection
 * and synchronous, so a plain db.transaction() gives the same atomicity
 * Python's explicit BEGIN IMMEDIATE provided. */
export function recordFailure(db: DbClient, username: string): LockoutStatus {
  const result = db.transaction((tx) => {
    const now = Date.now() / 1000
    const r = row(tx, username)
    const wasLocked = Boolean(r?.lockedUntil && r.lockedUntil > now)

    if (wasLocked) {
      return {
        locked: true,
        lockedUntil: r!.lockedUntil,
        failCount: r!.failCount,
        justLocked: false,
      }
    }

    let failCount: number
    let firstFailAt: number
    if (!r || !r.firstFailAt || now - r.firstFailAt > WINDOW_S) {
      failCount = 1
      firstFailAt = now
    } else {
      failCount = r.failCount + 1
      firstFailAt = r.firstFailAt
    }

    const lockedUntil = failCount >= THRESHOLD ? now + LOCKOUT_S : null

    tx.insert(loginAttempts)
      .values({ username, failCount, firstFailAt, lastFailAt: now, lockedUntil })
      .onConflictDoUpdate({
        target: loginAttempts.username,
        set: { failCount, firstFailAt, lastFailAt: now, lockedUntil },
      })
      .run()

    return {
      locked: lockedUntil !== null,
      lockedUntil,
      failCount,
      justLocked: lockedUntil !== null,
    }
  })

  if (result.justLocked) {
    // Fresh transition (wasLocked was false) -- log once, not every attempt.
    logEvent(
      db,
      'warn',
      `password login locked for ${(LOCKOUT_S / 60).toFixed(0)} min after ` +
        `${result.failCount} failed attempts (username=${JSON.stringify(username)})`,
    )
  }

  return {
    locked: result.locked,
    lockedUntil: result.lockedUntil,
    failCount: result.failCount,
  }
}

export function reset(db: DbClient, username: string): void {
  db.delete(loginAttempts).where(eq(loginAttempts.username, username)).run()
}
